buffer.go

```diff
@@ -0,0 +1,196 @@
+package erasurecode
+
+import (
+	"bytes"
+	"io"
+)
+
+type BufferMatrix struct {
+	b                []byte
+	hdrSize, bufSize int
+	len              int // len of input
+	k                int
+	curBlock         int
+	leftInBlock      int
+	finished         bool
+}
+
+// FragLen returns the size of a "fragment" aligned to a block size (data + header)
+func (b BufferMatrix) FragLen() int {
+	return b.SubGroups() * (b.bufSize + b.hdrSize)
+}
+
+// SubGroups returns the number of blocks inside a single fragment
+func (b BufferMatrix) SubGroups() int {
+	nbBlocks := (b.len + b.bufSize - 1) / b.bufSize
+	nbStripes := (nbBlocks + b.k - 1) / b.k
+	return nbStripes
+}
+
+func (b BufferMatrix) maxLen() int {
+	return (b.SubGroups() * b.k) * (b.bufSize + b.hdrSize)
+}
+
+// NewBufferMatrix returns a new buffer suitable for <len> data and organized
+// such as it can be injected into EncodeMatrixWithBuffer without allocation/copying
+// the data into shards
+func NewBufferMatrix(bufSize int, len int, k int) *BufferMatrix {
+	var b BufferMatrix
+	b.Reset(bufSize, len, k)
+	return &b
+}
+
+// Reset serves the same purpose as NewBufferMatrix but use the existing buffer and
+// tries to avoid allocation of the underlying buffer.
+func (b *BufferMatrix) Reset(bufSize int, len int, k int) {
+	hdrSize := fragmentHeaderSize()
+	b.hdrSize = hdrSize
+	b.bufSize = bufSize
+	b.len = len
+	b.k = k
+	b.leftInBlock = -1
+	b.curBlock = 0
+	b.finished = false
+
+	maxLen := b.maxLen()
+
+	if cap(b.b) <= maxLen {
+		// No internal buffer or buffer is too small, let's replace it
+		b.b, _ = memalign(maxLen, defaultAlign)
+	}
+	b.b = b.b[:]
+}
+
+var hdrBuffer = bytes.Repeat([]byte{0}, fragmentHeaderSize())
+
+// Finish *must* be called after the final Write() *before* using the buffer
+// in EncodeMatrix
+// It is safe to call it multiple times.
+func (b *BufferMatrix) Finish() {
+	if b.finished {
+		return
+	}
+	maxBlock := b.SubGroups() * b.k
+
+	for b.curBlock < maxBlock {
+		realCurBlock := b.getRealBlock(b.curBlock)
+		blockSize := b.hdrSize + b.bufSize
+		if b.leftInBlock == -1 {
+			// START OF BLOCK
+			copy(b.b[realCurBlock*(blockSize):], hdrBuffer)
```

Comment:
`realCurBlock * (blockSize)` is used twice, it could be computed once, and we use the result for both, but this is very minor changes.